### DISCUSSION TITLE
how to create a self host telegram server

### DISCUSSION THREAD

**Discussion by bitcard** [xelaj/mtproto#81] [Q&A] [answered]
This great project implement the whole mtproto protocol, and connect to office telegram server. how to create a telegram server based this project, is there any project to do this?

**Comment by quenbyako** [comment, marked as answer]
Hey @bitcard!

Unfortunately, server and client implementations are far-far different. The closest one project wht you want is [nebula chat](https://github.com/nebula-chat) **but here is a list of warns why NOT TO use mtproto anywhere except telegram:**

* the protocol is a shit spiced up with a spoon of crutches. seriously, look at current implementation, it is a whole agony at every single line! not cause contributors are script kiddies, but cause telegram developers are.
* even if you like it, there is a much simple way: you can just write your own customrpc+cbor+noise+coap implementation, and it could be twice faster and intuitive as mtproto! i'm not kidding. (we have one, but it's closed source yet, unfortunately) Mtproto is "yet another super secure but not amazing rpc protocol" which has more cons than pros.
* yeeeeah, telegram clients are amazing (and this is why telegram is the best messenger — great usability!), but note that **current tg clients cannot connect to other servers except telegram official ones.** This happend, cause every tg client hardcoded telegram's rsa key (xelaj/mtproto does not, but you can't use api without these rsa keys).

So, this is a multiple reasons why-not-to use mtproto in literaly any project except telegram clients. I warned you 😂
